crypto/fipsmodule/bn/rsaz_exp_x2.c

```diff
@@ -0,0 +1,623 @@
+    /* Expanded exponent */
+    BN_ULONG *expz = NULL;      /* [2][exp_digits + 1] */
+
+    /* Dual AMM */
+    DAMM damm = NULL;
+    /* Extractor from red_table */
+    DEXTRACT extract = NULL;
+
+/*
+ * Squaring is done using multiplication now. That can be a subject of
+ * optimization in future.
+ */
+# define DAMS(r,a,m,k0) damm((r),(a),(a),(m),(k0))
+
+    switch (modulus_bitsize) {
+    case 1024:
+        red_digits = 20;
+        exp_digits = 16;
+        damm = ossl_rsaz_amm52x20_x2_ifma256;
+        extract = ossl_extract_multiplier_2x20_win5;
+        break;
+    case 1536:
+        // Extended with 2 digits padding to avoid mask ops in high YMM register 
+        red_digits = 30 + 2;
+        exp_digits = 24;
+        damm = ossl_rsaz_amm52x30_x2_ifma256;
+        extract = ossl_extract_multiplier_2x30_win5;
+        break;
+    case 2048:
+        red_digits = 40;
+        exp_digits = 32;
+        damm = ossl_rsaz_amm52x40_x2_ifma256;
+        extract = ossl_extract_multiplier_2x40_win5;
+        break;
+    default:
+        goto err;
+    }
+
+    storage_len_bytes = (2 * red_digits                         /* red_Y     */
+                       + 2 * red_digits                         /* red_X     */
+                       + 2 * red_digits * (1U << exp_win_size)  /* red_table */
+                       + 2 * (exp_digits + 1))                  /* expz      */
+                       * sizeof(BN_ULONG)
+                       + 64;                                    /* alignment */
+
+    storage = (BN_ULONG *)OPENSSL_malloc(storage_len_bytes);
+    if (storage == NULL)
+        goto err;
+    OPENSSL_cleanse(storage, storage_len_bytes);
+    storage_aligned = (BN_ULONG *)align_pointer(storage, 64);
+
+    red_Y     = storage_aligned;
+    red_X     = red_Y + 2 * red_digits;
+    red_table = red_X + 2 * red_digits;
+    expz      = red_table + 2 * red_digits * (1U << exp_win_size);
+
+    /*
+     * Compute table of powers base^i, i = 0, ..., (2^EXP_WIN_SIZE) - 1
+     *   table[0] = mont(x^0) = mont(1)
+     *   table[1] = mont(x^1) = mont(x)
+     */
+    red_X[0 * red_digits] = 1;
+    red_X[1 * red_digits] = 1;
+    damm(&red_table[0 * 2 * red_digits], (const BN_ULONG*)red_X, rr, m, k0);
+    damm(&red_table[1 * 2 * red_digits], base,  rr, m, k0);
+
+    for (idx = 1; idx < (int)((1U << exp_win_size) / 2); idx++) {
+        DAMS(&red_table[(2 * idx + 0) * 2 * red_digits],
+             &red_table[(1 * idx)     * 2 * red_digits], m, k0);
+        damm(&red_table[(2 * idx + 1) * 2 * red_digits],
+             &red_table[(2 * idx)     * 2 * red_digits],
+             &red_table[1 * 2 * red_digits], m, k0);
+    }
+
+    /* Copy and expand exponents */
+    memcpy(&expz[0 * (exp_digits + 1)], exp[0], exp_digits * sizeof(BN_ULONG));
+    expz[1 * (exp_digits + 1) - 1] = 0;
+    memcpy(&expz[1 * (exp_digits + 1)], exp[1], exp_digits * sizeof(BN_ULONG));
+    expz[2 * (exp_digits + 1) - 1] = 0;
+
+    /* Exponentiation */
+    {
```

Comment:
I'm completely lost in this code. Can we please add proper documentation that either explains the whole algorithm or points to an algorithm in a paper where it's taken from?

Reply:
I am leaning more and more towards that there _has_ to be a more lucid way to do this, but this should be a follow up. For now, I've added a long comment at the top of the loop that explains the algorithm and which parts of the code correspond to what. I'll also drop it here:

```c
    // Exponentiation
    //
    // This is Algorithm 3 in iacr 2011-239 which is cited below as
    // well.
    //
    // Rather than compute base^{exp} in one shot, the powers of
    // base^i for i = [0..2^{exp_win_size}) are precomputed and stored
    // in `red_table`. Each window of the exponent is then used as an
    // index to look up the power in the table, and then that result
    // goes through a "series of squaring", which repositions it with
    // respect to where it appears in the complete exponent. That
    // result is then multiplied by the previous result.
    //
    // The `extract` routine does the lookup, `DAMS` wraps the `damm`
    // routine to set up squaring, while `damm` is the AMM
    // routine. That is what you find happening in each iteration of
    // this loop—stepping through the exponent one
    // `win_exp_size`-bit window at a time.
```